### PR TITLE
Code splitting and HMR

### LINKED ISSUE #221: Experiment: ESModule: A future without webpack
# Browser supporting
## Background
Chromium: ✔
Firefox: ✔
Firefox for Android: ❔
Firefox Geckoview: ❔
iOS: 👀 Can be done. Depends on `webextension-polyfill`

## Content 
Chromium: ✔
Firefox: ❌ https://bugzilla.mozilla.org/show_bug.cgi?id=1536094
Firefox for Android: ❌❔
Firefox Geckoview: ❌❔
iOS:  👀 Can be done. Depends on `webextension-polyfill`. Related https://github.com/Agoric/realms-shim/issues/47

## Options page:
Should not be a problem

## Pop up page:
Should not be a problem

# Toolchain supporting
📦 @pika/web https://github.com/pikapkg/web ❔ (not tested yet)
🔨 TypeScript custom transformation ✔

# Codebase that relies on Webpack
❔

### LINKED ISSUE #1609: [Tooling] Code-splitting & HMR
Investigate other's work on code-splitting, hot-module-reload in this region.

### PR DESCRIPTION
**Pull request by Jack-Works**
close #1609

close #372 close #221 (my conclusion is that the ecosystem is not ready for it. I need to re-invent too many tools for it)

cc @septs you can disable hmr by an environment variable "NO_HMR"

tested on Chrome (dev & prod) and Firefox (prod)

### CODE REVIEW

File: scripts/SSRPlugin.ts [hunk 1]
```diff
@@ -1,54 +1,32 @@
...
-    /**
-     * @returns {Promise<string>}
-     */
-    renderSSR() {
-        return new Promise((resolve, reject) => {
+export class SSRPlugin {
+    constructor(public htmlFileName: string, public pathName: string) {}
+    renderSSR(): Promise<string> {
+        return new Promise<string>((resolve, reject) => {
             exec('node ./src/setup.ssr.js ' + this.pathName, (err, stdout) => (err ? reject(err) : resolve(stdout)))
```
Comment (septs) [inline scripts/SSRPlugin.ts:8]:
```ts
import { execFile } from 'child_process';
import { promisify } from 'util';

const { stdout } = await promisify(execFile)('node', [this.pathName]);
return stdout
```

File: scripts/SSRPlugin.ts [hunk 2]
```diff
@@ -1,54 +1,32 @@
...
         })
     }
-    /**
-     * @param {string} original
-     * @param {string} string
-     */
-    appendAfterBody(original, string) {
+    appendAfterBody(original: string, string: string) {
```
Comment (septs) [inline scripts/SSRPlugin.ts:11]:
```suggestion
    appendAfterBody(original: string, prepare: string) {
```

File: scripts/SSRPlugin.ts [hunk 3]
```diff
@@ -1,54 +1,32 @@
...
         return original.replace('</body>', string + '</body>')
     }
-    /**
-     * @param {string} string
-     */
-    removeScripts(string) {
+    removeScripts(string: string) {
```
Comment (septs) [inline scripts/SSRPlugin.ts:14]:
```suggestion
    removeScripts(input: string) {
```

File: webpack.config.ts [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        Chromium: !!argv.chromium,
+        Safari: !!argv['wk-webview'],
+        E2E: !!argv.e2e,
+        Beta: !!argv.beta,
+        Insider: !!argv.insider,
+        ReproducibleBuild: !!argv['reproducible-build'],
+    }
+}
+/** Get git info */
+function getGitInfo() {
```
Comment (septs) [inline webpack.config.ts:306]: i think git info can generate a json file on release (only firefox)

File: webpack.config.ts [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+        REMOTE_URL: 'N/A',
+        BRANCH_NAME: 'N/A',
+        DIRTY: false,
+        TAG_DIRTY: false,
+    }
+}
+
+function toArray(x: string | string[]) {
+    return typeof x === 'string' ? [x] : x
+}
```
Comment (septs) [inline webpack.config.ts:334]:
```suggestion
function toArray(x: string | string[]) {
    return typeof x === 'string' ? [x] : x
}
```

move to https://github.com/DimensionDev/kit ?

File: webpack.config.ts (hunk 4 above)
Comment (Jack-Works) [reply webpack.config.ts:306]: good idea, plz open an issue for it, it's beyond of this pr

File: webpack.config.ts (hunk 5 above)
Comment (Jack-Works) [reply webpack.config.ts:334]: nope

File: webpack.config.ts (hunk 4 above)
Comment (septs) [reply webpack.config.ts:306]: fuck gh-bot issue title checking logic

File: scripts/SSRPlugin.ts [hunk 6]
```diff
@@ -1,54 +1,31 @@
...
+        return (await promisify(exec)('node ./src/setup.ssr.js ' + this.pathName)).stdout
     }
-    /**
-     * @param {string} original
-     * @param {string} string
-     */
-    appendAfterBody(original, string) {
-        return original.replace('</body>', string + '</body>')
+    appendAfterBody(original: string, text: string) {
+        return original.replace('</body>', text + '</body>')
```
Comment (guanbinrui) [inline scripts/SSRPlugin.ts:11]:
```suggestion
        return original.replace(`</body>${text}</body>`)
```

File: scripts/SSRPlugin.ts [hunk 7]
```diff
@@ -1,54 +1,31 @@
...
-     * @returns {Promise<string>}
-     */
-    renderSSR() {
-        return new Promise((resolve, reject) => {
-            exec('node ./src/setup.ssr.js ' + this.pathName, (err, stdout) => (err ? reject(err) : resolve(stdout)))
-        })
+export class SSRPlugin {
+    constructor(public htmlFileName: string, public pathName: string) {}
+    async renderSSR(): Promise<string> {
+        return (await promisify(exec)('node ./src/setup.ssr.js ' + this.pathName)).stdout
```
Comment (guanbinrui) [inline scripts/SSRPlugin.ts:8]:
```suggestion
        return (await promisify(exec)(`node ./src/setup.ssr.js ${this.pathName}`)).stdout
```

File: scripts/SSRPlugin.ts [hunk 8]
```diff
@@ -110,5 +84,12 @@ untilDocumentReady().then(() => {
         return regeneratedHTML
     }
 }
-
-module.exports = SSRPlugin
+function appendScript(html: string, src: string, position: string, position2: string) {
```
Comment (guanbinrui) [inline scripts/SSRPlugin.ts:87]: Plz rename`postion` and `position2` to better name. Maybe `position => tagName` and `position2 => position`?

**Review by guanbinrui — DISMISSED** [review]

**Pull request merged**